README.md
# Not with my name! Inferring artists' names of input strings employed by Diffusion Models

This is the official repository of  
  
**Not with my name! Inferring artists' names of input strings employed by Diffusion Models**. *[Roberto Leotta](https://www.linkedin.com/in/roberto-leotta-b68571176/)*, *[Oliver Giudice](https://www.linkedin.com/in/oliver-giudice/)*, *[Luca Guarnera](https://www.dmi.unict.it/lguarnera/)*, *[Sebastiano Battiato](https://www.dmi.unict.it/~battiato/)*. ICIAP 2023.  


[![arxiv](https://shields.io/badge/paper-green?logo=arxiv&style=for-the-badge)](https://arxiv.org/abs/2307.13527) [![webpage](https://shields.io/badge/Webpage-green?logo=safari&style=for-the-badge)](https://iplab.dmi.unict.it/mfs/Deepfakes/PaperArtistDeepfake-Iciap2023/)

![Not with my name!](docs/overall_output.png)

---

## Installation

Use [conda](https://conda.io/projects/conda/en/latest/user-guide/install/index.html) or [miniconda](https://docs.conda.io/en/latest/miniconda.html#) to easily install the required dependencies. 

With conda installed, run the following commands:

```bash
# 1. Clone the repository
git clone https://github.com/ictlab-unict/not-with-my-name.git
# 2. navigate to the repository folder
cd not-with-my-name
# 3. create the environment
conda env create -f docs/environment.yml
```

Then follow the instructions in the [dataset section](#dataset) and [checkpoints section](#checkpoints) to download the required files.

### Dataset

Download at least one of the following datasets:

| Dataset        | # Original images |  # AI generated images |      Link      |  GB  |
|:---------------|:-----------------:|:----------------------:|:--------------:|:----:|
| Small          |        2350       |          2350          | [link](https://drive.google.com/file/d/1HraHV0_yLd2fcQqY9k3bF7-yioRKyCob/view?usp=sharing)       |  2.7 |
| Medium         |        4130       |          8519          | _coming soon_  |   /  |

After downloading the dataset, extract it in the `resources` folder. The folder structure should be as the [following](resources/README.md).

### Checkpoints

Download the following checkpoint: [link](https://drive.google.com/file/d/1m8SVte9B7XbnFN_oRS_tZQm22WSkwEN8/view?usp=sharing) and place it in `resources/ckpts` folder. The folder structure should be as the [following](resources/README.md).

---

## Inference

__N.B.__: to run the inference you need to download the checkpoints and the dataset (at least the small one).

### Inference on a single image

To run the inference on a single image, run the following command:

```bash
# 1. activate the environment
conda activate not-w-my-name-env
# 2. run inference
python src/inference.py --dataset-folder <path-to-dataset> --query-img <path-to-query-image> --model-ckpt <path-to-checkpoint> --cuda --results-folder <path-to-results-folder>
```

__Results__ will be saved in the `--results-folder <path-to-results-folder>`.

#### Inference on a single image - example

```bash
# 1. activate the environment
conda activate not-w-my-name-env
# 2. run inference
python src/inference.py --dataset-folder resources/small-dataset/ --query-img resources/small-dataset/pablo_picasso/ai_generated/102_0.png --model-ckpt resources/ckpts/siamese_not_w_my_name.ckpt --cuda --results-folder results
```

### Inference usage

```
usage: inference.py [-h] [--show-time] [--debug] 
                    --dataset-folder DATASET_FOLDER --results-folder RESULTS_FOLDER 
                    --query-img QUERY_IMAGE [--distance-th DISTANCE_TH] [--cuda]
                    --model-ckpt MODEL_CKPT

Not with my name inference by Roberto Leotta

optional arguments:
  -h, --help            show this help message and exit
  --show-time           show processing time
  --debug               flag for development debugging
  --dataset-folder DATASET_FOLDER
                        dataset folder path
  --results-folder RESULTS_FOLDER
                        results folder path
  --query-img QUERY_IMAGE
                        query image path
  --distance-th DISTANCE_TH
                        distance threshold for the query image. Default: 0.5
  --cuda                use CUDA for inference
  --model-ckpt MODEL_CKPT
                        siamese model checkpoint
```

---

## Citation

If you find this code useful for your research, please cite our paper:

```
@inproceedings{leotta2023not,
   title={Not with my name! Inferring artists' names of input strings employed by Diffusion Models},
   author={Leotta, Roberto and Giudice, Oliver and Guarnera, Luca and Battiato, Sebastiano},
   booktitle={International Conference on Image Analysis and Processing},
   year={2023},
   organization = {Springer}
}
```

---

## Credits
__Authors__: [Roberto Leotta](https://www.linkedin.com/in/roberto-leotta-b68571176/), [Oliver Giudice](https://www.linkedin.com/in/oliver-giudice/), [Luca Guarnera](https://www.dmi.unict.it/lguarnera/), [Sebastiano Battiato](https://www.dmi.unict.it/~battiato/)  
__Version__: 1.0.1  
__Date__: 08/22/2023

CHANGELOG.md
# CHANGELOG

## [1.0.1] - 08-22-2023

### Minor changes 

- Fixing README.md typos

## [1.0.0] - 08-21-2023

- First release

## [0.1] - 07-27-2023

- Updating README.md with paper link

## [0.0] - 07-14-2023

- Repo creation

src/utils.py
'''
    @Author: Roberto Leotta
    @Version: 1.0
    @Date: 08/16/2023

    Utils snippets.
'''
########## IMPORTs START ##########

# system imports
import os
import sys

# processing
import cv2
import numpy as np

###### IMPORTs - END ##############

###
###
###

###### CONSTANTs - START ##########

ROOT_DIR = os.path.abspath(os.path.join(os.path.realpath(__file__), '..', '..'))
ALLOWED_FILES_EXT = ['jpg', 'png']

###### CONSTANTs - END ############

###
###
###

###### FUNCTIONs - START ##########

def p_info(message, n_tabs=0, end='\n', pre=''):
    """
    Function to print an info message.
    :param message: message to print in stdout
    :param n_tabs: initial blank tab(s). Default n_tabs=0
    :param end: end character for each printed message. Default end='\n'
    :param pre: prefix to prepend to the message. Default pre=''
    """
    if message == '':
        print('**')
    else:
        tabs_str = ''
        for tab in range(0, n_tabs):
            tabs_str += '\t'
        print('{}** (I) ** {}{}'.format(pre, tabs_str, message), end=end)


def p_warning(message, n_tabs=0, end='\n', pre=''):
    """
    Function to print an warning message.
    :param message: message to print in stdout
    :param n_tabs: initial blank tab(s). Default n_tabs=0
    :param end: end character for each printed message. Default end='\n'
    :param pre: prefix to prepend to the message. Default pre=''
    """
    tabs_str = ''
    for tab in range(0, n_tabs):
        tabs_str += '\t'
    print('{}** (W) ** {}{}'.format(pre, tabs_str, message), end=end)


def p_error_n_exit(message, n_tabs=0):
    """
    Function to print an error message and exit.
    :param message: message to print in stdout
    :param n_tabs: initial blank tab(s). Default n_tabs=0
    """
    tabs_str = ''
    for tab in range(0, n_tabs):
        tabs_str += '\t'
    print('** (E) **{}- {}'.format(tabs_str, message))

    sys.exit(1)


def p_error(message, n_tabs=0):
    """
    Function to print an error message, without exit.
    :param message: message to print in stdout
    :param n_tabs: initial blank tab(s). Default n_tabs=0
    """
    if message == '':
        print('**')
    else:
        tabs_str = ''
        for tab in range(0, n_tabs):
            tabs_str += '\t'
        print('** (E) ** {}{}'.format(tabs_str, message))


def str2bool(v):
    """
    Convert string to boolean.
    :param v: string to convert in boolean
    :return bool: string converted in True or False
    """
    return v.lower() in ("yes", "y", "true", "t", "1")


def allowedFile(file_path):
    """
    Function to allow the right files.
    :param file_path: file_path
    :return bool: True if allowed, False otherwise
    """
    return True if file_path.split('.')[-1] in ALLOWED_FILES_EXT else False


def show(imgs, name='Results'):
    """
    Function to show image, or multiple images (vertical stucked)
    :param imgs: image or images as list. i.e.: ((img)) or ((img1, img2))
    :param name: name of the window to show
    """
    if imgs.__class__ != tuple:
        res = imgs
    else:
        res = np.vstack((imgs))

    cv2.namedWindow(name, cv2.WINDOW_NORMAL)
    cv2.resizeWindow(name, res.shape[1], res.shape[0])
    cv2.imshow(name, res)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

###### FUNCTIONs - END ############

###
###
###

###### COMMENTs ###################

src/SiameseNet.py
'''
    Siamese network.
'''
########## IMPORTs START ##########

# processing
import cv2

# nn
import torch
import torch.optim as optim
from torch.nn import functional as F
import pytorch_lightning as pl
import warnings
warnings.filterwarnings("ignore")

# data augmentation
import albumentations as A
from albumentations.pytorch import ToTensorV2

# utils
import utils

###### IMPORTs - END ##############

###
###
###

###### CONSTANTs - START ##########

test_transforms = A.Compose(
    [
        A.SmallestMaxSize(max_size=512),
        A.CenterCrop(height=512, width=512),
        A.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
        ToTensorV2(),
    ]
)

###### CONSTANTs - END ############

###
###
###

###### CLASSEs - START ##########

class SiameseNetworkTask(pl.LightningModule):
    def __init__(self, 
                embedding_net,          # embedding network
                _optimizer='sgd',       # optimizer
                lr=0.01,                # learning rate
                momentum=0.99,          # momentum
                margin=2,               # margin loss
                weight_decay=0.0001     # weight decay
            ):
        super(SiameseNetworkTask, self).__init__()
        self.save_hyperparameters()
        self.embedding_net = embedding_net
        # loss definition
        self.criterion = ContrastiveLoss(margin)
        self._optimizer = _optimizer
                    
    def forward(self, x):
        return self.model(x)
        
    def configure_optimizers(self):
        if self._optimizer == 'sgd':
            return optim.SGD(self.embedding_net.parameters(), 
                                self.hparams.lr, 
                                momentum=self.hparams.momentum)
        elif self._optimizer == 'adam':
            return optim.Adam(self.embedding_net.parameters(), 
                                lr=self.hparams.lr, 
                                weight_decay=self.hparams.weight_decay)
        else:
            utils.p_error_n_exit('Optimizer not supported: {}'.format(self._optimizer))
    
    def training_step(self, batch, batch_idx):
        # take into account the I_i and I_j images and the label l_ij
        I_i, I_j, l_ij, *_ = batch
        
        # compute the embeddings
        phi_i = self.embedding_net(I_i)
        phi_j = self.embedding_net(I_j)

        # loss computation
        l, _ = self.criterion(phi_i, phi_j, l_ij)
        
        self.log('train/loss', l)
        return l
    
    def validation_step(self, batch, batch_idx):
        I_i, I_j, l_ij, *_ = batch
        phi_i = self.embedding_net(I_i)
        phi_j = self.embedding_net(I_j)
        l, _ = self.criterion(phi_i, phi_j, l_ij)
        self.log('valid/loss', l)
        
        if batch_idx==0:
            self.logger.experiment.add_embedding(phi_i, batch[3], I_i, global_step=self.global_step)


class ContrastiveLoss(torch.nn.Module):
    def __init__(self, m=2):
        super(ContrastiveLoss, self).__init__()
        self.m = m

    def forward(self, phi_i, phi_j, l_ij):
        d = F.pairwise_distance(phi_i, phi_j)
        l = 0.5 * (1 - l_ij.float()) * torch.pow(d, 2) + \
            0.5 * l_ij.float() * torch.pow( torch.clamp( self.m - d, min = 0) , 2)
        return l.mean(), d
    
###### CLASSEs - END ############

###
###
###

###### FUNCTIONs - START ##########

def compute_embedding(image_filepath, model, device):
    """
    Function to compute the embedding of an image.
    :param image_filepath: path of the image
    :param model: model to use
    :param device: device to use
    :return: embedding
    """
    image = cv2.imread(image_filepath)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    image = test_transforms(image=image)["image"]

    # get embedding
    x = image.unsqueeze(0).to(device)
    embedding = model(x)
    embedding = embedding.detach().cpu()

    return embedding

###### FUNCTIONs - END ############

###
###
###

###### COMMENTs ###################

src/inference.py
'''
    @Author: Roberto Leotta
    @Version: 1.0
    @Date: 08/16/2023

    env:
    conda activate not-w-my-name-env

    How to run:
    python src/inference.py --dataset-folder resources/small-dataset/ --query-img resources/small-dataset/pablo_picasso/ai_generated/102_0.png --model-ckpt resources/ckpts/siamese_not_w_my_name.ckpt --cuda --results-folder results
'''
########## IMPORTs START ##########

# system imports
import os
import time
import argparse

# nn
import torch
from SiameseNet import SiameseNetworkTask, compute_embedding
import torch.nn.functional as F

# utils
import utils
import pandas as pd
from tqdm import tqdm

###### IMPORTs - END ##############

###
###
###

###### CONSTANTs - START ##########

###### CONSTANTs - END ############

###
###
###

###### FUNCTIONs - START ##########

###### FUNCTIONs - END ############

###
###
###

###### MAIN - START ###############

if __name__ == "__main__":    
    ### 
    # initial settings

    parser = argparse.ArgumentParser(description='Not with my name inference\nby Roberto Leotta')
    parser.add_argument('--show-time', dest='show_time', default=True, action='store_true', 
                        help='show processing time')
    parser.add_argument('--debug', required=False, default=False, action='store_true', 
                        help='flag for development debugging')
    
    parser.add_argument('--dataset-folder', dest='dataset_folder', type=str, required=True, 
                        help='dataset folder path')
    parser.add_argument('--results-folder', dest='results_folder', required=True, default='', type=str, 
                        help='results folder path')
    parser.add_argument('--query-img', dest='query_image', type=str, required=True, 
                        help='query image path')
    parser.add_argument('--distance-th', dest='distance_th', type=float, required=False, default=0.5,
                        help='distance threshold for the query image. Default: 0.5')
    
    parser.add_argument('--cuda', required=False, default=False, action='store_true',
                        help='use CUDA for inference')
    parser.add_argument('--model-ckpt', dest='model_ckpt', required=True, type=str, 
                        help='siamese model checkpoint')
    args = parser.parse_args()

    # check if input folder exists
    DATASET_FOLDER = os.path.abspath(args.dataset_folder)
    if not os.path.exists(DATASET_FOLDER):
        utils.p_error_n_exit('Dataset folder does not exist: {}'.format(DATASET_FOLDER))

    # check if model checkpoint exists
    MODEL_CKPT = os.path.abspath(args.model_ckpt)
    if not os.path.exists(MODEL_CKPT):
        utils.p_error_n_exit('Model checkpoint does not exist: {}'.format(MODEL_CKPT))

    # check if query image exists
    QUERY_IMAGE = os.path.abspath(args.query_image)
    if not os.path.exists(QUERY_IMAGE):
        utils.p_error_n_exit('Query image does not exist: {}'.format(QUERY_IMAGE))

    # create results folder
    if args.results_folder != '':
        RESULTS_FOLDER = os.path.abspath(args.results_folder)
    else:
        RESULTS_FOLDER = os.path.join(utils.ROOT_DIR, 'results')
    os.makedirs(RESULTS_FOLDER, exist_ok=True)

    # save and print the used config
    with open(os.path.join(RESULTS_FOLDER, 'config.txt'), 'w') as config_file:
        config_file.write('*** CONFIG ***\n')
        for arg in args.__dict__:
            if args.__dict__[arg] is not None:
                config_file.write(str(arg) + ': ' + str(args.__dict__[arg]) + '\n')

    print('*** CONFIG ***\n', args, '\n')

    if args.show_time:
        print('*** START ***\n')
        start_time = time.time()

    ###
    # init device
    if args.cuda and torch.cuda.is_available():
        DEVICE = torch.device('cuda')
    else:
        DEVICE = torch.device('cpu')
    # print used device
    utils.p_info('Using device: {}'.format(DEVICE))

    # init model
    model = SiameseNetworkTask.load_from_checkpoint(MODEL_CKPT, map_location=DEVICE)
    model.eval()
    if args.debug: utils.p_info('Network:\n{}'.format(model)); utils.p_info('')
    utils.p_info('Siamese network loaded from checkpoint: {}'.format(MODEL_CKPT))

    ###
    # create a dataframe with the original images and the relative embeddings;
    # the dataframe will be used to compute the distance between the query image and the original images.
    utils.p_info('Computing embeddings for original images and generated images...')
    original_imgs_list = []
    generated_imgs_list = []
    for root, dirs, files in os.walk(DATASET_FOLDER):
        if files != [] and 'original_paintings' in root:
            total_file = len(files)
        else: continue
        with tqdm(total=total_file, desc='Processing {} embeddings'.format(os.path.basename(os.path.dirname(root)) + '/' + os.path.basename(root))) as pbar:
            for file in files:
                if utils.allowedFile(file):
                    # retrieve the image path, the class name and the original/generated flag
                    img_path = os.path.join(root, file)
                    class_name = os.path.basename(os.path.dirname(root))
                    original_generated = os.path.basename(root)
                    original_generated = 'original' if original_generated == 'original_paintings' else 'generated'
                    # compute the embedding
                    embedding = compute_embedding(img_path, model.embedding_net, DEVICE)

                    if original_generated == 'original':
                        original_imgs_list.append([img_path, class_name, original_generated, embedding])
                    else:
                        generated_imgs_list.append([img_path, class_name, original_generated, embedding])

                    pbar.update(1)

    # create the dataframe
    df_original_imgs = pd.DataFrame(original_imgs_list, columns=['img_path', 'class', 'original_generated', 'embedding'])
    # free up memory
    del original_imgs_list

    # compute the embeddings for the query image
    utils.p_info('Computing embedding for query image...')
    query_embedding = compute_embedding(QUERY_IMAGE, model.embedding_net, DEVICE)

    # add a new column to the dataframe with the distance between the query image and the original images
    df_original_imgs['query_distance'] = None

    # compute the distance
    utils.p_info('Computing distance between query image and original images...')
    for index, row in df_original_imgs.iterrows():
        orginal_embedding = row['embedding']
        distance = F.pairwise_distance(orginal_embedding, query_embedding, p=2).detach().cpu().numpy()[0]

        if distance <= args.distance_th:
            # save the distance
            df_original_imgs.at[index, 'query_distance'] = distance
        else:
            # remove the row in wich the distance is greater than the threshold
            df_original_imgs.drop(index, inplace=True)

    # sort the dataframe by distance
    df_original_imgs = df_original_imgs.sort_values(by=['query_distance'])
    # remove the embedding column
    df_original_imgs = df_original_imgs.drop(columns=['embedding'])
    # save the dataframe
    utils.p_info('Saving results in {}'.format(os.path.join(RESULTS_FOLDER, 'distances.csv')))
    df_original_imgs.to_csv(os.path.join(RESULTS_FOLDER, 'distances.csv'), index=False)

    ###
    # exit

    if args.show_time: print('\n*** END in {:0.2f} secs ***\n'.format(time.time() - start_time))


###### MAIN - END #################

###
###
###

###### COMMENTs ###################